### PR TITLE
Interactive Visualization

### PR DESCRIPTION
**Pull request by jakspacek**
## Proposed changes
Add plotly interactive graphs to the Validator page.

## Notes
I have a good idea of the roadmap to fully integrate these, but I don't have the resources right now to test the changes properly myself.
- the routine `generate_visualizations()` (from generate_visualizations.py) should be called in controller.py's `validator()` (around line 405, maybe?)
- `generate_visualizations()` is equipped to dump working HTML to a file (specified by an argument) - it will be wrapped in a <body> etc but those are easy to trim out. It's approximately 3MB of javascript.
- Either `render_template()` (the call on line 406 of controller.py currently) can be modified to take another argument similar to the 'result' string, or you can staple the figures to 'result' itself for a convenient way to test without changing any architecture?

## Types of changes
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionalitiy)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (if none of the other choices apply)

## Checklist
- [ ] Changes have been tested to ensure that fix is effective or that a feature works.
- [ ] Changes pass the unit tests
- [x] I have included necessary documentation or comments (as necessary)
- [ ] Any dependent changes have been merged and published

_PR template was adopted from [appium](https://github.com/appium/appium/blob/master/.github/PULL_REQUEST_TEMPLATE.md)_

### CODE REVIEW

File: generate_visualizations.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            if cpy < 0:
+                out = interval * j
+                break
+        #
+        outformatted = str(round(out,2)) + "-" + str(round(out + interval,2))
+        output.append(outformatted)
+    #
+    return output
+
+generate_visualizations()
```
Comment (tkkuehn) [inline generate_visualizations.py:187]: This is okay for testing, but should remove once this is integrated into the main validator

**Review by tkkuehn — COMMENTED** [review]
Really nice work! The output graphs look super nice. I have a few comments on this as is, which I'll put here for reference, then address before moving on to integrating this with the existing site.

**Comment by tkkuehn** [comment]
Alrighty, I think this is in shape for a review now.

**Review by kaitj — COMMENTED** [review]
Code wise, it looks good to me - have not had a chance to look at it locally yet (maybe a quick demo on the next call?). 

One thought, though it doesn't have to be changed for this PR, for the `generate_visualizations`, wondering if it might be a good idea to have separate functions / more meaningful names in case more figures are to be added in the future?

**Comment by tkkuehn** [comment]
Notes from today's meeting:

- Put table back
- Label colorbar
- Remove tiny text on histogram

**Review by kaitj — CHANGES_REQUESTED** [review]
Clean up code (see meeting minutes & @tkkuehn's comments)

**Comment by tkkuehn** [comment]
I think this is ready for another review!

**Review by kaitj — COMMENTED** [review]
Looks great, both code and actual site. Changes look good to me, and good to merge. Will note some of the code will likely need to be changed with the creation of an afids class (#101). Wondering do we want to merge it into master now, or release it as part of 0.2.x?

**Comment by tkkuehn** [comment]
> Wondering do we want to merge it into master now, or release it as part of 0.2.x?

Either way seems fine to me. We could even merge this, make it 0.2.0, then make the AFIDs class stuff 0.3.0.

**Comment by kaitj** [comment]
Given that I somehow did some weird thing with creating the AFIDs class and this doesn't actually break functionality with 0.1.x (though it is a major enhancement), I'd be happy with merging this in as 0.1.x and still call the AFIDs class 0.2.x

**Comment by tkkuehn** [comment]
Great, feel free to merge and tag it as a new release in that case!

**Review by kaitj — APPROVED** [review]

**Pull request merged**